### ISSUE TITLE
/setpassword sends password in plain text

### ISSUE THREAD

**Issue by tuedel** [minetest/minetest#8982]
##### Minetest version
```
5.1.0-dev
```
@ https://github.com/minetest/minetest/commit/d77ea764768f5cde9e8f0781356a67d723aa7609 (Current HEAD)

##### Summary
For the `/setpassword` chat command to be processed, it currently needs to be sent to the server in plain text, including the newly set password.

The password should only be exchanged using SRP, so the chat command should be handled by the client somehow.

**Comment by sfan5** [comment]
`/setpassword` is only meant to be used for administrative purposes like resetting a password to a temporary one.
Clients can already change their own password using SRP.

**Comment by tuedel** [comment]
IMHO it should at least be mentioned somewhere that `/setpassword` is inherently insecure, both for sending the new password in plain text and storing it as a SHA-1 hash.

Ideally, the already existing method for changing a password using SRP could be used on other players by those with the `password` privilege.

Is this a feature request then? ;)

**Comment by 56independent** [comment]
```
@@ -416,7 +416,7 @@ core.register_chatcommand("revokeme", {
 
 core.register_chatcommand("setpassword", {
 	params = S("<name> <password>"),
-	description = S("Set player's password"),
+	description = S("Set player's password (sent unencrypted, thus insecure)"),
 	privs = {password=true},
 	func = function(name, param)
 		local toname, raw_password = string.match(param, "^([^ ]+) +(.+)$")
```

I can't seem to git push yet, even with my access token for some reason. 

Attached is the diff

**Comment by JazzGlobal** [comment]
> ```
> @@ -416,7 +416,7 @@ core.register_chatcommand("revokeme", {
>  
>  core.register_chatcommand("setpassword", {
>  	params = S("<name> <password>"),
> -	description = S("Set player's password"),
> +	description = S("Set player's password (sent unencrypted, thus insecure)"),
>  	privs = {password=true},
>  	func = function(name, param)
>  		local toname, raw_password = string.match(param, "^([^ ]+) +(.+)$")
> ```
> 
> I can't seem to git push yet, even with my access token for some reason.
> 
> Attached is the diff

Try forking the repo, creating a branch on your fork, and trying to push to that branch.

**Comment by 56independent** [comment]
I can't push anywhere due to a credential error as Minetest is hosted on Github, a Microsoft platform. Microsoft is the leading manufacturer of broken products.
I did make a branch but not a fork; i never came to realise they are different lol.

**Comment by JazzGlobal** [comment]
> I can't push anywhere due to a credential error as Minetest is hosted on Github, a Microsoft platform. Microsoft is the leading manufacturer of broken products.
> I did make a branch but not a fork; i never came to realise they are different lol.

Not so sure it's an issue with Microsoft. If you aren't the repo owner and haven't been given permission to push to the repo then you need to fork it, create a branch from your fork, push to that branch and open a pull request to merge your branch (from your fork) to the main code base. This is just how that works. 

Regardless, I'm glad you were able to figure it out.

**Comment by appgurueu** [comment]
First of all: This issue is way too narrow in its scope. The underlying issue is the lack of network encryption, which applies to all chatcommands - heck, all interactions of players with the game, be they through formspecs, nodes, entities, whatever even! Just patching `/setpassword` would still leave all other commands vulnerable, which is about as problematic considering commands like WE's `//lua` or the builtin `/clearpassword`, `/set`, `/grant`, `/revoke`, `/remove_player`, `/rollback`, `/shutdown` etc., all of which can be used to wreak havoc.

Now let's consider possible attack scenarios: To exploit the `/setpassword` vulnerability - in order to steal a password - you basically need a man-in-the-middle (MITM) attack: You need to listen in on the connection between admin/moderator and server. This would allow you to read the chat message (chat command), which is transmitted in plain text, allowing you to obtain the password, which could then be used to log in as the player.

Now, if you already are a MITM, you know the admin/moderator's IP & peer ID. You are thus able to *spoof arbitrary packets* in the name of the admin/moderator (I assume), allowing you to **invoke `/setpassword` and other sensitive commands yourself as you please.** You don't need to wait for the admin/moderator to do this.

In particular, this means https://github.com/minetest/minetest/pull/13049/ is practically absolutely useless: It fails entirely to get rid of the more severe issue at play - the ability for a MITM attacker to spoof commands. **`/setpassword` does not need to be invoked for the vulnerability to come into play.**

It also is way too narrow: The plaintext warning applies to all chatcommands. `/setpassword` is not the only sensitive chatcommand.

The only proper fix for this issue is https://github.com/minetest/minetest/issues/10206. I believe this issue should be closed in favor of the network encryption issue.

**Comment by JazzGlobal** [comment]
> First of all: This issue is way too narrow in its scope. The underlying issue is the lack of network encryption, which applies to all chatcommands - heck, all interactions of players with the game, be they through formspecs, nodes, entities, whatever even! Just patching `/setpassword` would still leave all other commands vulnerable, which is about as problematic considering commands like WE's `//lua` or the builtin `/clearpassword`, `/set`, `/grant`, `/revoke`, `/remove_player`, `/rollback`, `/shutdown` etc., all of which can be used to wreak havoc.
> 
> Now let's consider possible attack scenarios: To exploit the `/setpassword` vulnerability - in order to steal a password - you basically need a man-in-the-middle (MITM) attack: You need to listen in on the connection between admin/moderator and server. This would allow you to read the chat message (chat command), which is transmitted in plain text, allowing you to obtain the password, which could then be used to log in as the player.
> 
> Now, if you already are a MITM, you know the admin/moderator's IP & peer ID. You are thus able to _spoof arbitrary packets_ in the name of the admin/moderator (I assume), allowing you to **invoke `/setpassword` and other sensitive commands yourself as you please.** You don't need to wait for the admin/moderator to do this.
> 
> In particular, this means #13049 is practically absolutely useless: It fails entirely to get rid of the more severe issue at play - the ability for a MITM attacker to spoof commands. **`/setpassword` does not need to be invoked for the vulnerability to come into play.**
> 
> It also is way too narrow: The plaintext warning applies to all chatcommands. `/setpassword` is not the only sensitive chatcommand.
> 
> The only proper fix for this issue is #10206. I believe this issue should be closed in favor of the network encryption issue.

Thorough explanation. This issue should be closed then.

**Comment by tuedel** [comment]
> The only proper fix for this issue is #10206. I believe this issue should be closed in favor of the network encryption issue.
I concur that this would be the ideal solution.

**Comment by appgurueu** [comment]
@minetest/triagers should be able to take care of this.
